tune-threads/tune-threads.lisp

```diff
@@ -0,0 +1,92 @@
+  "Time the quil program in qvm Q and return the number of seconds."
+  (qvm::prepare-for-parallelization num-threads)
+  (simple-time num-trials (run-and-reset-qvm q)))
+
+(defun time-quil-prog (q &key (num-trials 1) (num-threads 4))
+  "Time the quil program in qvm Q and return the number of seconds per thread per trial."
+  (/ (time-quil-prog-raw q :num-trials num-trials :num-threads num-threads) (* num-threads num-trials)))
+
+(defun find-num-trials (q num-threads time-limit &optional (hint 0))
+  "Choose the number of trials for the qvm/program Q that take around TIME-LIMIT seconds to complete."
+  (if (= hint 1)
+      1
+      (loop :for num-trials := 1 :then (* num-trials 2)
+            :for one-time := 0.0 :then (time-quil-prog-raw q :num-threads num-threads :num-trials num-trials)
+            :while (< one-time 0.1) ; avoid divide by zero and times very close to zero
+            :finally (return (1+ (floor (/ time-limit (/ one-time num-trials))))))))
+
+(defun hadamard-program-source (n)
+  "Return a quil source program that performs a Hadamard gate on each of N qubits and measures each qubit."
+  (with-output-to-string (s)
+    (format s "DECLARE ro BIT[~d]~%" n)
+    (dotimes (q n)
+      (format s "H ~d~%" q))
+    (dotimes (q n)
+      (format s "MEASURE ~d ro[~d]~%" q q))))
+
+(defun hadamard-program (n)
+  "Return a parsed quil program that performs a Hadamard gate on each of N qubits and measures each qubit."
+  (let ((cl-quil::*allow-unresolved-applications* t))
+    (quil:parse-quil (hadamard-program-source n))))
+
+(defun prepare-hadamard-test (num-qubits)
+  "Return a qvm loaded with a program created by HADAMARD-PROGRAM with NUM-QUBITS qubits."
+  (let* ((q (qvm:make-qvm num-qubits))
+         (p (hadamard-program num-qubits)))
+    (qvm:load-program q p :supersede-memory-subsystem t)
+    q))
+
+(defun time-hadamard-raw (&key (num-trials 1) (num-threads 4) (num-qubits 10))
+  (time-quil-prog-raw (prepare-hadamard-test num-qubits) :num-trials num-trials :num-threads num-threads))
+
+(defun time-hadamard (&key (num-trials 1) (num-threads 4) (num-qubits 10))
+  "Time the Hadamard program and return the number of seconds per thread per trial."
+  (time-quil-prog (prepare-hadamard-test num-qubits) :num-trials num-trials :num-threads num-threads))
+
+(defun scan-num-threads (&key (time-limit 5.0) (num-qubits 12)
+                              (max-num-threads (qvm:count-logical-cores))
+                              (min-num-threads 1)
+                              (q (prepare-hadamard-test num-qubits)))
+  "Return a list of execution time per sample per thread of the qvm/program Q for different numbers of threads.
+The number of threads varies from MIN-NUM-THREADS to MAX-NUM-THREADS."
+  (norm-min
+   (loop :for num-trials := (find-num-trials q num-threads time-limit 0)
+           :then (find-num-trials q num-threads time-limit num-trials)
+         :for num-threads :from min-num-threads :to max-num-threads
```

Comment:
I think you might want to put this `:for num-threads` clause first, since the expression that initializes `num-trials` references `num-threads`?  I believe the bindings happen sequentially, as if in a `let*`.